several improvements

shadiakiki1986 opened the pull request:
Hello. Thanks for this package. It saved me a lot of time. I made some improvements in this PR. Here's a summary:
* documentation: pre-requisites for installation, note about which mongodb/collection class this mocks
* implement the function for collection name
* add support for option 'upsert'
* add support for atomic modifiers: set, unset
* remove a possible code line duplicate about bson document

martin-helmich commented:
Awesome, thanks for your effort! I'll have a detailed look at this shortly (and also investigate why the unit tests are failing -- which I suspect to be unrelated to this change).

martin-helmich commented on composer.json:
```diff
@@ -29,6 +28,13 @@
     }
   },
   "require-dev": {
+    "phpunit/phpunit": "^5.2",
```
Not sure about this; many of the actual (not test) classes depend on PHPUnit classes, so PHPUnit is actually a hard dependency of this package (which would typically be included as a dev dependency itself).

shadiakiki1986 replied:
You're right. I rebased the repo and just kept the added test script in the composer.json file ([commit here](https://github.com/martin-helmich/php-mongomock/pull/1/commits/8fc6924fc94b088608bec35b048c440e3661b199))

shadiakiki1986 replied:
You're right. I edited that commit and kept the added test script in the composer json file

martin-helmich commented on src/MockCollection.php:
```diff
@@ -49,7 +49,14 @@ public function insertOne($document, array $options = [])
+        // e.g. $document->bsonSerialize() will give the same result
+        // so I couldn't write a test to capture the need or not for this.
+        // I'm commenting it out anyway.
+        // $document = new BSONDocument($document);
```
Good catch; you actually _can_ initialize a `BSONDocument` with another one, and it will automatically unbox the value. It doesn't hurt, but also doesn't do any good. I'd be fine with removing this statement altogether.

martin-helmich commented on src/MockCollection.php:
```diff
@@ -80,24 +87,60 @@ public function updateOne($filter, $update, array $options = [])
-            foreach ($update['$set'] ?? [] as $k => $v) {
-                $doc[$k] = $v;
+            $this->updateCore($doc,$update); 
+            $anyUpdates = $anyUpdates || true;
```
I don't get this. _"[something] or true"_ should always be _true_. `$anyUpdates = true` should have the same effect.

shadiakiki1986 replied:
yea this is just an artifact of the fixup. I had originally written `$anyUpdates = $anyUpdates || !!$update['$set']`, but then from the docs I realized that in case of `!!$update['$set']`, the element just gets completely replaced (check [line 128](https://github.com/martin-helmich/php-mongomock/pull/1/files#diff-b0c966a257e74dfc0b81dff64b6cf9adR128))

shadiakiki1986 replied:
You're right. I edited that commit as suggested

martin-helmich commented on src/MockCollection.php:
```diff
@@ -80,24 +87,60 @@ public function updateOne($filter, $update, array $options = [])
+            $doc[$k] = $v;
+        }
+
+        if (array_key_exists('$unset',$update)) {
```
Technically, the `array_key_exists` check is superflous; the `??` operator in the nextline also contains an implicit `isset` check, so this would work one way or the other.

shadiakiki1986 replied:
You're right. I edited that commit and removed the check

martin-helmich commented:
@shadiakiki1986, I've added some feedback to your changes. Most issues are more a matter of opinion, but I'd prefer to keep `phpunit/phpunit` as a regular dependency. What are your thoughts on this?

Also, after merging #2, the unit tests are now also passing (tested locally).

shadiakiki1986 commented:
Hi Martin. I'm glad you found the PR of use. I accidentally ran a `git push -f` from a copy of my fork that was 11 commits behind while rebasing. Then I found another copy of my fork which I thought was up-to-date, but that was also 4 commits behind. (so sloppy of me on a friday evening) Anyway, I have the up-to-date fork, but it is not within my reach now. I'll get to it tomorrow, run a `git push -f` to get back on track, and then integrate fixes for your comments.

shadiakiki1986 commented:
Ok... back in business after an accidental `git push -f` with an outdated version of my fork (phew). I found a copy that had the complete set of commits in this PR before my blunder. Then I rebased again, integrated your comments, and also squashed the commits having to do with adding features to updateMany/updateOne. Let me know if they look good

martin-helmich commented:
:+1: from me! Thanks for the contribution!
